atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraResources.java

```diff
@@ -0,0 +1,167 @@
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.keyvalue.cassandra;
+
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.concurrent.Callable;
+
+import org.junit.rules.ExternalResource;
+import org.junit.rules.RuleChain;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import com.jayway.awaitility.Awaitility;
+import com.jayway.awaitility.Duration;
+import com.palantir.atlasdb.cassandra.CassandraKeyValueServiceConfigManager;
+import com.palantir.atlasdb.cassandra.ImmutableCassandraCredentialsConfig;
+import com.palantir.atlasdb.cassandra.ImmutableCassandraKeyValueServiceConfig;
+import com.palantir.atlasdb.config.ImmutableLeaderConfig;
+import com.palantir.atlasdb.config.LeaderConfig;
+import com.palantir.atlasdb.ete.Gradle;
+import com.palantir.docker.compose.DockerComposeRule;
+import com.palantir.docker.compose.connection.DockerPort;
+import com.palantir.docker.compose.connection.waiting.HealthCheck;
+import com.palantir.docker.compose.connection.waiting.SuccessOrFailure;
+
+class CassandraResources extends ExternalResource {
+    private final Logger log = LoggerFactory.getLogger(CassandraResources.class);
+    private final int THRIFT_PORT_NUMBER = 9160;
+    private final Gradle GRADLE_PREPARE_TASK = Gradle.ensureTaskHasRun(
+            ":atlasdb-ete-test-utils:buildCassandraImage");
+    private final DockerComposeRule docker = DockerComposeRule.builder()
+            .file("src/test/resources/docker-compose.yml")
+            .waitingForHostNetworkedPort(THRIFT_PORT_NUMBER, toBeOpen())
+            .saveLogsTo("container-logs")
+            .build();
+    private final RuleChain CASSANDRA_DOCKER_SET_UP = RuleChain.outerRule(GRADLE_PREPARE_TASK).around(docker);
+
+    private static int refCount = 0;
+    private static CassandraResources currentInstance;
+
+    static InetSocketAddress CASSANDRA_THRIFT_ADDRESS;
+    static ImmutableCassandraKeyValueServiceConfig CASSANDRA_KVS_CONFIG;
+    static Optional<LeaderConfig> LEADER_CONFIG;
+
+    static synchronized CassandraResources getCassandraResource() {
+        if (refCount == 0) {
+            // currentInstance either hasn't been created yet, or after was called on it - create a new one
+            currentInstance = new CassandraResources();
+        }
+        return currentInstance;
+    }
+
+    private CassandraResources() {
+        // nothing to initialize
+    }
+
+    @Override
+    protected void before() {
+        synchronized (CassandraResources.class) {
+            try {
+                if (refCount == 0) {
+                    try {
+                        waitUntilCassandraIsUp();
+                    } catch (Throwable e) {
+                        log.error("Exception: {}", e);
```

Comment:
If Cassandra never comes up, we should fail fast / throw immediately instead of just logging it (that way, it's easier to figure out why the test failed). 

Also, not convinced refCount should increase in that case.